feat(blog): post about aggregation queries (#37)
someengineering/resoto.com: blog/resoto-query-language-aggregation/index.md

The simplest way of aggregating data using Resoto is the `count` command. It allows for counting objects or counting the occurrences of a specific property. Let's assume we are interested in the number of compute instances we maintain. We can get the answer by executing the following command:

```shell
> query is(instance) | count
total matched: 540
total unmatched: 0
```

All compute instances for all cloud providers are handled as `instance` in Resoto, so `is(instance)` will select them (see [aws_ec2_instance](/docs/reference/data-models/aws#aws_ec2_instance) or [gcp_instance](/docs/reference/data-models/gcp#gcp_instance)). The `count` command takes this list of instances and counts the number of occurrences. The `count` command also allows specifying a property to use as a grouping value. The following query would show the instances grouped by instance status. Each compute instance has a reported property named `instance_status`. The number of occurrences of this property value is counted here:

```shell

> > query is(instance) | count instance_status

Author: I replaced shell->bash everywhere.
Can you explain why this is better? (Ideally we could write resotoshell...)

stopped: 48
terminated: 151
running: 341
total matched: 540
total unmatched: 0
```

While `count` is easy to use for many cases, it is not powerful enough to do advanced aggregations. Let's assume we want to know the available CPU cores and memory of all instances in our clouds. We can achieve this by using the `aggregate` function. See the following example:

> ```shell
> > query is(instance) | aggregate 

Reviewer: ```suggestion
```bash
$> query is(instance) | aggregate 
```

  sum(instance_cores) as sum_of_cores,
  max(instance_cores) as max_cores, 
  sum(instance_memory) as sum_of_memory, 
  max(instance_memory) as max_mem
sum_of_cores: 3441
max_cores: 16
sum_of_memory: 12802.25
max_mem: 64
```

As you can see, we have 3441 cores in total, while there is no instance with more than 16 cores. The same data is also available for the provisioned memory: almost 13TB of RAM, while no instance has more than 64GB of RAM. We could have also queried using `min` or `avg` functions.

While this aggregated data is already useful, we would like to break it down using grouping variables. We have already seen grouping variables in the `count` command: a group is defined by the value of a specified property. Let's assume we want to aggregate available memory broken down by instance status.

> ```shell
> > query is(instance) | aggregate 

Reviewer: ```suggestion
```bash
$> query is(instance) | aggregate 
```

   instance_status as status:
   sum(instance_memory) as memory
group:
 status: running
memory: 8538
---
group:
 status: stopped
memory: 1345
---
group:
 status: terminated
memory: 2919.25
```

> This query will create multiple results, where each result has a group property. In this group object, you will find the values of the grouping variables. The grouping functions are all applied to the instances in this group. So we see compute instances running with a sum of 8TB, while the remaining 4TB are stopped or terminated.

Reviewer: ```suggestion

This query will create multiple results, where each result has a group property. In this group object, you will find the values of the grouping variables. The grouping functions are all applied to the instances in this group. So we see compute instances running with a sum of 8TB, while the remaining 4TB are stopped or terminated.
```

## A tale about ancestors and descendants

> The aggregation capabilities we have seen so far allow for grouping and functions that get applied to the result of each group. Since we are dealing with a graph, it would be great if we could aggregate not only the data of a single node but over the data of nodes that are linked in the graph! RQL allows performing merge queries to allow precisely this: they are mighty and deserve their own little blog post. But there is a handy short notation for merge queries that we want to use here, that already covers a lot of use cases:

Author: You are right. I tried to rephrase the sentence. PTAL.

![Instance Relationships](./img/carz_structure.svg)

> The diagram shows the relationship between instances. AWS resources are attached to a region, while GCP resources link to a zone. Every compute instance also has a related `instance_type` node as a predecessor in the graph. To access properties of ancestor nodes of a given kind relative to a node, we can use the `ancestors` notation:
> 
> Example:
> ```shell
> > query is(instance) | aggregate 

Reviewer: ```suggestion
The diagram shows the relationship between instances. AWS resources are attached to a region, while GCP resources link to a zone. Every compute instance also has a related `instance_type` node as a predecessor in the graph. To access properties of ancestor nodes of a given kind relative to a node, we can use the `ancestors` notation:

```bash
$> query is(instance) | aggregate 
```

  sum(/ancestors.instance_type.reported.ondemand_cost) as cost
cost: 155.73
```

> This query filters all instances and then aggregates the on-demand cost of each element by walking the graph up to the instance type and selecting the `reported.ondemand_cost` property from this node. The path `/ancestors.instance_type.reported.ondemand_cost` can be translated into a walk over the node's ancestors until one of kind `instance_type` is found. The walk itself is not defined explicitly, but added automagically by Resoto. The remaining part of this path is relative to the node that is found, which is `reported.ondemand_cost` in this example. The result of this aggregation query is the on-demand cost of all instances that we maintain.

Reviewer: ```suggestion

This query filters all instances and then aggregates the on-demand cost of each element by walking the graph up to the instance type and selecting the `reported.ondemand_cost` property from this node. The path `/ancestors.instance_type.reported.ondemand_cost` can be translated into a walk over the node's ancestors until one of kind `instance_type` is found. The walk itself is not defined explicitly, but added automagically by Resoto. The remaining part of this path is relative to the node that is found, which is `reported.ondemand_cost` in this example. The result of this aggregation query is the on-demand cost of all instances that we maintain.
```

It is possible to walk the graph inbound with `ancestors` and outbound via `descendants`. Let's use this technique to filter only the running instances and aggregate the data using account and region. In a query where you can write a property path, you can also use the `ancestors` syntax - aggregation groups and functions included:

> ```shell
> > query is(instance) and instance_status==running | aggregate 

Reviewer: ```suggestion
```bash
$> query is(instance) and instance_status==running | aggregate 
```

tags: [query language, aws, gcp]
---

# Aggregating Data with Resoto

:::note

> This blog post is the second in a series about Resoto's powerful query language. The previous post, [Resoto Query Language 101](/blog/2022/02/04/resoto-query-language-101), provides an introduction to the query language.

Reviewer: ```suggestion
This blog post is the second in a series about Resoto's powerful search functionality. The previous post, [Resoto Search 101](/blog/2022/02/04/resoto-search-101), provides an introduction to Resoto's search capabilities.
```

Reviewer: You'll want to rebase this branch to `main` so that Docusaurus doesn't think this is a broken link (the build will fail on broken internal links).

:::

> Resoto's query language allows selecting resources using filters, [combinators](/docs/reference/cli/query/basic-queries#combining-selections), and [traversals](/docs/reference/cli/query/basic-queries#selecting-nodes-by-traversal). Query results can be [combined](/docs/reference/cli/query/basic-queries#combining-selections), grouped, and [aggregated](/docs/reference/cli/query/aggregation).

Reviewer: ```suggestion
Resoto's query language allows selecting resources using [filters](/docs/concepts/search/filters), [combinators](/docs/concepts/search/filters#combining-selections), and [traversals](/docs/concepts/search/traversals). Query results can be [combined](/docs/concepts/search/filters#combining-selections), grouped, and [aggregated](/docs/concepts/search/aggregation).
```
